### PR TITLE
APP-1179/APP-1180 : IB APIs to search for Jira users / assign new ticket to user

### PR DESCRIPTION
**Pull request by Hamitay-Daitan**

### CODE REVIEW

File: src/main/java/org/symphonyoss/integration/jira/api/JiraApiResource.java [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+
+import javax.ws.rs.core.MediaType;
+
+/**
+ * REST endpoint to handle requests for JIRA Api.
+ *
+ * Created by alexandre-silva-daitan on 08/08/17.
+ */
+@RestController
+@RequestMapping("/v1/{configurationId}/rest/api")
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/api/JiraApiResource.java]: Please change it to: "@RequestMapping("/v1/jira/rest/api")"

File: src/main/java/org/symphonyoss/integration/jira/api/JiraApiResource.java [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+
+  private static final String PATH_JIRA_API_SEARCH_USERS =
+      "rest/api/latest/user/assignable/search?issueKey=%s&username=%s&maxResults=%s";
+
+  private static final String PATH_JIRA_API_ASSIGN_ISSUE =
+      "/rest/api/latest/issue/%s/assignee";
+
+  private static final String COMPONENT = "JIRA API";
+
+  private final IntegrationBridge integrationBridge;
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/api/JiraApiResource.java]: I think it won't required anymore. Change this attribute to JiraWebHookIntegration class and fix the constructor injection.

File: src/main/java/org/symphonyoss/integration/jira/api/JiraApiResource.java [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+
+  private final LogMessageSource logMessage;
+
+  private final JwtAuthentication jwtAuthentication;
+
+  private final UserAssignService userAssignService;
+
+  private final SearchAssignableUsersService searchAssignableUsersService;
+
+  private final Integer maxResults = new Integer(10);
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/api/JiraApiResource.java]: Hardcoded is not a good solution. We should modify it to a parameterized value on the YAML config file and keep the 10 value as a fallback.

File: src/main/java/org/symphonyoss/integration/jira/api/JiraApiResource.java [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+
+  private String accessToken;
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/api/JiraApiResource.java]: It's used on the method scope, it's not required to be an attribute.

File: src/main/java/org/symphonyoss/integration/jira/api/JiraApiResource.java [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+      SearchAssignableUsersService searchAssignableUsersService) {
+    this.integrationBridge = integrationBridge;
+    this.logMessage = logMessage;
+    this.jwtAuthentication = jwtAuthentication;
+    this.userAssignService = userAssignService;
+    this.searchAssignableUsersService = searchAssignableUsersService;
+  }
+
+  /**
+   * Get a list of potential assigneers users from an especific Issue.
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/api/JiraApiResource.java]: Get a list of potential assignable users from a specific issue.

File: src/main/java/org/symphonyoss/integration/jira/api/JiraApiResource.java [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+   * @param issueKey Issue identifier
+   * @param username User that made a request from JIRA
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/api/JiraApiResource.java]: The username you want to query from JIRA

File: src/main/java/org/symphonyoss/integration/jira/api/JiraApiResource.java [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+   * @return List of potential assigneers users or 400 Bad Request - Returned if no issue key
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/api/JiraApiResource.java]: @return List of potential assignable users or 
   400 Bad Request - Returned if no issue key was provided or invalid
   401 Unauthorized - Returned if the user is not authenticated

File: src/main/java/org/symphonyoss/integration/jira/api/JiraApiResource.java [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+   * was provided, 401 Unauthorized - Returned if the user is not authenticated ,
+   * 404 Not Found - Returned if the requested user is not found.
+   */
+  @GetMapping(value = "/user/assignable/search", produces = MediaType.APPLICATION_JSON)
+  public ResponseEntity searchAssignableUsers(@RequestParam String issueKey,
+      @RequestParam(required = false) String username, @PathVariable String configurationId,
+      @RequestHeader(value = "Authorization", required = false) String authorizationHeader,
+      @RequestParam(name = "url") String jiraIntegrationURL)
+      throws IOException {
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/api/JiraApiResource.java]: Why IOException?

File: src/main/java/org/symphonyoss/integration/jira/api/JiraApiResource.java [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+
+    Long userId = jwtAuthentication.getUserIdFromAuthorizationHeader(authorizationHeader);
+    AuthorizedIntegration authIntegration = getAuthorizedIntegration(configurationId);
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/api/JiraApiResource.java]: Change this to: validateIntegrationBootstrap(jiraWebHookIntegration)

File: src/main/java/org/symphonyoss/integration/jira/api/JiraApiResource.java [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+    try {
+      provider = authIntegration.getOAuth1Provider(jiraIntegrationURL);
+    } catch (OAuth1Exception e) {
+      throw new IntegrationRuntimeException(COMPONENT,
+          logMessage.getMessage(APPLICATION_KEY_ERROR), e);
+    }
+    return provider;
+  }
+
+  private String getAccessToken(
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/api/JiraApiResource.java]: Once this method has only one line I think it's not required

File: src/main/java/org/symphonyoss/integration/jira/api/JiraApiResourceConstants.java [hunk 11]
```diff
@@ -0,0 +1,11 @@
+package org.symphonyoss.integration.jira.api;
+
+/**
+ * Created by hamitay on 8/15/17.
+ */
+public class JiraApiResourceConstants {
+
+  public static String ISSUE_KEY = "issueKey";
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/api/JiraApiResourceConstants.java]: It should be 'public static final'.

File: src/main/java/org/symphonyoss/integration/jira/api/JiraApiResourceConstants.java [hunk 12]
```diff
@@ -0,0 +1,11 @@
+package org.symphonyoss.integration.jira.api;
+
+/**
+ * Created by hamitay on 8/15/17.
+ */
+public class JiraApiResourceConstants {
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/api/JiraApiResourceConstants.java]: Missing javadoc.

File: src/main/java/org/symphonyoss/integration/jira/api/JiraApiResourceExceptionHandler.java [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+import org.symphonyoss.integration.authorization.AuthorizationException;
+import org.symphonyoss.integration.exception.IntegrationRuntimeException;
+import org.symphonyoss.integration.model.ErrorResponse;
+
+/**
+ * Created by hamitay on 8/15/17.
+ */
+
+@ControllerAdvice
+public class JiraApiResourceExceptionHandler {
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/api/JiraApiResourceExceptionHandler.java]: Missing javadoc.

File: src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java [hunk 14]
```diff
@@ -277,21 +278,15 @@ private String getCallbackUrl(IntegrationSettings settings) {
    * @throws AuthorizationException Thrown in case of error.
    */
   public boolean isUserAuthorized(IntegrationSettings settings, String url, Long userId)
-      throws AuthorizationException {
+      throws AuthorizationException, OAuth1HttpRequestException {
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java]: The OAuth1HttpRequestException exception won't be required anymore.

File: src/main/java/org/symphonyoss/integration/jira/properties/ServiceProperties.java [hunk 15]
```diff
@@ -0,0 +1,20 @@
+package org.symphonyoss.integration.jira.properties;
+
+/**
+ * Created by hamitay on 8/15/17.
+ */
+public class ServiceProperties {
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/properties/ServiceProperties.java]: Missing javadoc.

File: src/main/java/org/symphonyoss/integration/jira/services/SearchAssignableUsersService.java [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+import org.symphonyoss.integration.model.ErrorResponse;
+
+import java.io.IOException;
+import java.net.URL;
+
+/**
+ * Created by alexandre-silva-daitan on 15/08/17.
+ */
+@Component
+public class SearchAssignableUsersService {
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/services/SearchAssignableUsersService.java]: Missing javadoc.

File: src/main/java/org/symphonyoss/integration/jira/webhook/JiraWebHookIntegration.java [hunk 17]
```diff
@@ -138,7 +141,8 @@ public AppAuthorizationModel getAuthorizationModel() {
    * @see AuthorizedIntegration#isUserAuthorized(String, Long)
    */
   @Override
-  public boolean isUserAuthorized(String url, Long userId) throws AuthorizationException {
+  public boolean isUserAuthorized(String url, Long userId)
+      throws AuthorizationException, OAuth1HttpRequestException {
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/webhook/JiraWebHookIntegration.java]: The OAuth1HttpRequestException exception won't be required anymore.

File: src/test/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManagerTest.java [hunk 18]
```diff
@@ -206,7 +207,8 @@ public void testInvalidPublicKeyFile() throws URISyntaxException {
   }
 
   @Test(expected = NullPointerException.class)
-  public void testIsUserAuthorizedException() throws AuthorizationException, URISyntaxException {
+  public void testIsUserAuthorizedException()
+      throws AuthorizationException, URISyntaxException, OAuth1HttpRequestException {
```
Comment (robson-sanchez) [inline src/test/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManagerTest.java]: The OAuth1HttpRequestException exception won't be required anymore.

File: src/test/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManagerTest.java [hunk 19]
```diff
@@ -228,7 +230,8 @@ public void testIsUserAuthorizedException() throws AuthorizationException, URISy
   }
 
   @Test(expected = JiraOAuth1Exception.class)
-  public void testIsUserAuthorizedInvalidURL() throws AuthorizationException {
+  public void testIsUserAuthorizedInvalidURL()
+      throws AuthorizationException, OAuth1HttpRequestException {
```
Comment (robson-sanchez) [inline src/test/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManagerTest.java]: The OAuth1HttpRequestException exception won't be required anymore.

**Review by robson-sanchez — CHANGES_REQUESTED** [review]

File: src/main/java/org/symphonyoss/integration/jira/api/JiraApiResource.java (hunk 1 above)
Comment (ardsilva) [reply src/main/java/org/symphonyoss/integration/jira/api/JiraApiResource.java]: Done!
96199b287619b90657d6a5d4dee3224a1687b0fa

File: src/main/java/org/symphonyoss/integration/jira/api/JiraApiResource.java (hunk 2 above)
Comment (ardsilva) [reply src/main/java/org/symphonyoss/integration/jira/api/JiraApiResource.java]: Done!
79afb405f3a718c7cd037b56b3272b0c114a121a

File: src/main/java/org/symphonyoss/integration/jira/api/JiraApiResource.java (hunk 4 above)
Comment (ardsilva) [reply src/main/java/org/symphonyoss/integration/jira/api/JiraApiResource.java]: Done!
bc49f6e6bff398a7228c359488e73dc0097fd02d

File: src/main/java/org/symphonyoss/integration/jira/api/JiraApiResource.java (hunk 9 above)
Comment (ardsilva) [reply src/main/java/org/symphonyoss/integration/jira/api/JiraApiResource.java]: Done!
ca72590eb57fcd9832505185cf7a216795e3e6e8

File: src/main/java/org/symphonyoss/integration/jira/api/JiraApiResource.java (hunk 5 above)
Comment (ardsilva) [reply src/main/java/org/symphonyoss/integration/jira/api/JiraApiResource.java]: Done!
4cba9bee7500073e2f257b78ec47e1e799d93799

File: src/main/java/org/symphonyoss/integration/jira/api/JiraApiResource.java (hunk 6 above)
Comment (ardsilva) [reply src/main/java/org/symphonyoss/integration/jira/api/JiraApiResource.java]: Done!
cc98bf40c021f47559baa6eac98aee5cb871e6fd

File: src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java (hunk 14 above)
Comment (ardsilva) [reply src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java]: Done!
7bd958140d410badc01e5a63ebf9e42161ff0d52

File: src/main/java/org/symphonyoss/integration/jira/webhook/JiraWebHookIntegration.java (hunk 17 above)
Comment (ardsilva) [reply src/main/java/org/symphonyoss/integration/jira/webhook/JiraWebHookIntegration.java]: Done!
7bd958140d410badc01e5a63ebf9e42161ff0d52

File: src/test/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManagerTest.java (hunk 18 above)
Comment (ardsilva) [reply src/test/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManagerTest.java]: Done!
5f99cea017dec4bc48b3bc64022713e42f648a7b

File: src/test/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManagerTest.java (hunk 19 above)
Comment (ardsilva) [reply src/test/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManagerTest.java]: Done!
5f99cea017dec4bc48b3bc64022713e42f648a7b

File: src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java [hunk 20]
```diff
@@ -376,7 +373,7 @@ public void authorizeTemporaryToken(IntegrationSettings settings, String tempora
    * @return JiraOAuth1Provider configured.
    * @throws JiraOAuth1Exception Thrown in case of error.
    */
-  private JiraOAuth1Provider getJiraOAuth1Provider(IntegrationSettings settings, String baseUrl)
+  public JiraOAuth1Provider getJiraOAuth1Provider(IntegrationSettings settings, String baseUrl)
       throws JiraOAuth1Exception {
     AppAuthorizationModel appAuthorizationModel = getAuthorizationModel(settings);
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java:404]: We are reading public/private key on the filesystem on every request. It's a lot of unnecessary I/O operations.

File: src/main/java/org/symphonyoss/integration/jira/exception/InvalidJiraURLException.java [hunk 21]
```diff
@@ -0,0 +1,12 @@
+package org.symphonyoss.integration.jira.exception;
+
+/**
+ * Created by hamitay on 8/16/17.
+ */
+public class InvalidJiraURLException extends RuntimeException {
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/exception/InvalidJiraURLException.java]: Missing javadoc. All unchecked exceptions must extend IntegrationRuntimeException

**Review by robson-sanchez — CHANGES_REQUESTED** [review]

File: src/main/java/org/symphonyoss/integration/jira/services/SearchAssignableUsersService.java [hunk 22]
```diff
@@ -0,0 +1,55 @@
...
+
+  @Autowired
+  private LogMessageSource logMessage;
+
+  public ResponseEntity searchAssingablesUsers(String accessToken, OAuth1Provider provider,
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/services/SearchAssignableUsersService.java]: Missing javadoc.

You don't need to pass component as a parameter, this a constant. You could define it as a constant field in a separate class that holds the constant fields.

File: src/main/java/org/symphonyoss/integration/jira/services/SearchAssignableUsersService.java [hunk 23]
```diff
@@ -0,0 +1,55 @@
...
+      URL myselfUrl, String component, String issueKey) throws IOException {
+
+    if (issueKey == null || issueKey.isEmpty()) {
+      ErrorResponse response = new ErrorResponse();
+      response.setStatus(HttpStatus.BAD_REQUEST.value());
+      return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(response);
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/services/SearchAssignableUsersService.java]: Missing the message how to fix it.

File: src/main/java/org/symphonyoss/integration/jira/services/SearchAssignableUsersService.java [hunk 24]
```diff
@@ -0,0 +1,55 @@
...
+ * Created by alexandre-silva-daitan on 15/08/17.
+ */
+@Component
+public class SearchAssignableUsersService {
+
+  @Autowired
+  private LogMessageSource logMessage;
+
+  public ResponseEntity searchAssingablesUsers(String accessToken, OAuth1Provider provider,
+      URL myselfUrl, String component, String issueKey) throws IOException {
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/services/SearchAssignableUsersService.java]: IOException?

Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/services/SearchAssignableUsersService.java]: Change 'myselfUrl' to 'assignableUserUrl'.

File: src/main/java/org/symphonyoss/integration/jira/services/SearchAssignableUsersService.java [hunk 25]
```diff
@@ -0,0 +1,55 @@
...
+
+    if (issueKey == null || issueKey.isEmpty()) {
+      ErrorResponse response = new ErrorResponse();
+      response.setStatus(HttpStatus.BAD_REQUEST.value());
+      return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(response);
+    }
+
+    HttpResponse response = null;
+    try {
+      response = provider.makeAuthorizedRequest(accessToken, myselfUrl, HttpMethods.GET, null);
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/services/SearchAssignableUsersService.java]: You should change to:

```
HttpResponse response = provider.makeAuthorizedRequest(accessToken, myselfUrl, HttpMethods.GET, null);
return ResponseEntity.ok().body(response.parseAsString());
```

File: src/main/java/org/symphonyoss/integration/jira/services/SearchAssignableUsersService.java [hunk 26]
```diff
@@ -0,0 +1,55 @@
...
+    } catch (OAuth1HttpRequestException e) {
+      if (e.getCode() == HttpStatus.NOT_FOUND.value()) {
+        ErrorResponse errorResponse = new ErrorResponse();
+        errorResponse.setStatus(HttpStatus.NOT_FOUND.value());
+        errorResponse.setMessage(e.getLocalizedMessage());
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/services/SearchAssignableUsersService.java]: Are you sure that's the real message you should answer to the user?

**Review by robson-sanchez — CHANGES_REQUESTED** [review]

File: src/main/java/org/symphonyoss/integration/jira/services/SearchAssignableUsersService.java (hunk 24 above)
Comment (ardsilva) [reply src/main/java/org/symphonyoss/integration/jira/services/SearchAssignableUsersService.java]: Done!
801056d13c2165b996f1b73df969ffd789ec4d60

File: src/main/java/org/symphonyoss/integration/jira/api/JiraApiResourceConstants.java (hunk 12 above)
Comment (ardsilva) [reply src/main/java/org/symphonyoss/integration/jira/api/JiraApiResourceConstants.java]: Done!
f823bbfe5f1edafa89a71daaa18090b37aa6b81e

File: src/main/java/org/symphonyoss/integration/jira/api/JiraApiResourceConstants.java (hunk 11 above)
Comment (ardsilva) [reply src/main/java/org/symphonyoss/integration/jira/api/JiraApiResourceConstants.java]: Done!
c36918cbf64c4fe9b07ae0cb629ebd7b55c6f492

File: src/main/java/org/symphonyoss/integration/jira/api/JiraApiResourceExceptionHandler.java (hunk 13 above)
Comment (ardsilva) [reply src/main/java/org/symphonyoss/integration/jira/api/JiraApiResourceExceptionHandler.java]: done!
23e54a93288f8c88f745c5727c6328446f7f39c5

File: src/main/java/org/symphonyoss/integration/jira/properties/ServiceProperties.java (hunk 15 above)
Comment (ardsilva) [reply src/main/java/org/symphonyoss/integration/jira/properties/ServiceProperties.java]: done!
f3300cb5c1e9b5174216735c79b00e3418dc140a

File: src/main/java/org/symphonyoss/integration/jira/services/SearchAssignableUsersService.java (hunk 16 above)
Comment (ardsilva) [reply src/main/java/org/symphonyoss/integration/jira/services/SearchAssignableUsersService.java]: Done!
e872e578747710504dc6c3d6fe4f7e039fd199a6

File: src/main/java/org/symphonyoss/integration/jira/services/SearchAssignableUsersService.java (hunk 22 above)
Comment (ardsilva) [reply src/main/java/org/symphonyoss/integration/jira/services/SearchAssignableUsersService.java]: Done!
acb124f0616c1359cfa66cee37778a54a77c6aec

File: src/main/java/org/symphonyoss/integration/jira/services/SearchAssignableUsersService.java (hunk 24 above)
Comment (ardsilva) [reply src/main/java/org/symphonyoss/integration/jira/services/SearchAssignableUsersService.java]: done!
c1aecfc649b75febca9d3a17f2736e6d6313c0b8

File: src/main/java/org/symphonyoss/integration/jira/services/SearchAssignableUsersService.java (hunk 23 above)
Comment (ardsilva) [reply src/main/java/org/symphonyoss/integration/jira/services/SearchAssignableUsersService.java]: done!
49815acdde0fe1714adada8aacf57f325d3686ad

File: src/main/java/org/symphonyoss/integration/jira/services/SearchAssignableUsersService.java (hunk 25 above)
Comment (ardsilva) [reply src/main/java/org/symphonyoss/integration/jira/services/SearchAssignableUsersService.java]: done!
c4e55be16c602e4e7200b278f0be298999c09c43

File: src/main/java/org/symphonyoss/integration/jira/services/SearchAssignableUsersService.java (hunk 26 above)
Comment (ardsilva) [reply src/main/java/org/symphonyoss/integration/jira/services/SearchAssignableUsersService.java]: done!
c4e55be16c602e4e7200b278f0be298999c09c43

File: src/main/java/org/symphonyoss/integration/jira/services/SearchAssignableUsersService.java (hunk 22 above)
Comment (ardsilva) [reply src/main/java/org/symphonyoss/integration/jira/services/SearchAssignableUsersService.java]: 3f3a27738ac03f7dbe8f50e4aee047270c60076d

File: src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java [hunk 27]
```diff
@@ -305,11 +300,13 @@ public boolean isUserAuthorized(IntegrationSettings settings, String url, Long u
...
           provider.makeAuthorizedRequest(jiraOAuth1Data.getAccessToken(), myselfUrl,
               HttpMethods.GET, null);
-
       return response.getStatusCode() != HttpStatusCodes.STATUS_CODE_UNAUTHORIZED;
     } catch (MalformedURLException e) {
       throw new JiraOAuth1Exception(logMessage.getMessage("integration.jira.url.api.invalid", url),
           e, logMessage.getMessage("integration.jira.url.api.invalid.solution"));
+    } catch (OAuth1HttpRequestException e) {
+      throw new JiraOAuth1Exception(logMessage.getMessage("integration.jira.url.api.invalid", url),
+          e, logMessage.getMessage("integration.jira.url.api.invalid.solution"));
```
Comment (rsanchez-daitan) [inline src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java]: Change this message to: integration.jira.unreachable

integration.jira.unreachable = "Cannot access JIRA instance"
integration.jira.unreachable.solution = "Make sure the JIRA instance {0} is working fine"

File: src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java [hunk 28]
```diff
@@ -387,4 +384,38 @@ private JiraOAuth1Provider getJiraOAuth1Provider(IntegrationSettings settings, S
     provider.configure(consumerKey, privateKey, baseUrl, callbackUrl);
     return provider;
   }
+
+  public String getAccessToken(IntegrationSettings settings, String url, Long userId)
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java]: Missing javadoc

File: src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java [hunk 29]
```diff
@@ -387,4 +384,38 @@ private JiraOAuth1Provider getJiraOAuth1Provider(IntegrationSettings settings, S
...
+    JiraOAuth1Data jiraOAuth1Data = getJiraOAuth1Data(userAuthorizationData);
+
+    if (StringUtils.isEmpty(jiraOAuth1Data.getAccessToken())) {
+      return null;
+    }
+
+    return jiraOAuth1Data.getAccessToken();
+  }
+
+  private JiraOAuth1Data getJiraOAuth1Data(UserAuthorizationData userAuthorizationData)
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java]: Missing javadoc

File: src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java [hunk 30]
```diff
@@ -387,4 +384,38 @@ private JiraOAuth1Provider getJiraOAuth1Provider(IntegrationSettings settings, S
...
+
+    try {
+      jiraOAuth1Data = JsonUtils.readValue(userAuthorizationData.getData(), JiraOAuth1Data.class);
+    } catch (IOException e) {
+      throw new AuthorizationException("Invalid temporary token");
+    }
+    return jiraOAuth1Data;
+  }
+
+  public UserAuthorizationData getUserAuthorizationData(IntegrationSettings settings, String url,
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java]: Missing javadoc

File: src/main/java/org/symphonyoss/integration/jira/services/SearchAssignableUsersService.java [hunk 31]
```diff
@@ -0,0 +1,85 @@
...
+@Component
+public class SearchAssignableUsersService {
+
+  @Autowired
+  private LogMessageSource logMessage;
+
+  public ResponseEntity searchAssingablesUsers(String accessToken, OAuth1Provider provider,
+      URL assignableUserUrl, String issueKey) {
+
+    if (StringUtils.isEmpty(issueKey)) {
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/services/SearchAssignableUsersService.java]: Duplicated code. Throws a common exception and deal with this on the exception handler.

File: src/main/java/org/symphonyoss/integration/jira/services/SearchAssignableUsersService.java [hunk 32]
```diff
@@ -0,0 +1,85 @@
...
+      response.setMessage(logMessage.getMessage(MISSING_FIELD, ISSUE_KEY));
+      return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(response);
+    }
+
+    try {
+      HttpResponse response =
+          provider.makeAuthorizedRequest(accessToken, assignableUserUrl, HttpMethods.GET, null);
+      return ResponseEntity.ok().body(response.parseAsString());
+    } catch (OAuth1HttpRequestException e) {
+      if (e.getCode() == HttpStatus.NOT_FOUND.value()) {
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/services/SearchAssignableUsersService.java:52]: Duplicated code. Throws a common exception and deal with this on the exception handler.

File: src/main/java/org/symphonyoss/integration/jira/services/SearchAssignableUsersService.java [hunk 33]
```diff
@@ -0,0 +1,85 @@
...
+/**
+ * Service to get a list of potential assignable users from a specific issue.
+ *
+ * Created by alexandre-silva-daitan on 15/08/17.
+ */
+@Component
+public class SearchAssignableUsersService {
+
+  @Autowired
+  private LogMessageSource logMessage;
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/services/SearchAssignableUsersService.java]: Change the LogMessageSource to MessageUtils.

File: src/main/java/org/symphonyoss/integration/jira/services/UserAssignService.java [hunk 34]
```diff
@@ -0,0 +1,95 @@
...
+public class UserAssignService {
+
+  @Autowired
+  private LogMessageSource logMessage;
+
+  public ResponseEntity assignUserToIssue(String accessToken, String issueKey, String username,
+      URL integrationURL, OAuth1Provider provider) {
+
+    //Validate input
+    if (StringUtils.isEmpty(issueKey)) {
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/services/UserAssignService.java]: Duplicated code. Throws a common exception and deal with this on the exception handler.

File: src/main/java/org/symphonyoss/integration/jira/services/UserAssignService.java [hunk 35]
```diff
@@ -0,0 +1,95 @@
...
+ * Service that handles the assign user to issue on the JIRA API
+ * Created by hamitay on 8/15/17.
+ */
+@Component
+public class UserAssignService {
+
+  @Autowired
+  private LogMessageSource logMessage;
+
+  public ResponseEntity assignUserToIssue(String accessToken, String issueKey, String username,
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/services/UserAssignService.java]: Missing javadoc

File: src/main/java/org/symphonyoss/integration/jira/services/UserAssignService.java [hunk 36]
```diff
@@ -0,0 +1,95 @@
...
+    //Jira requisition
+    try {
+
+      GenericData data = new GenericData();
+      data.put(NAME_PATH, username);
+      JsonHttpContent content = new JsonHttpContent(new JacksonFactory(), data);
+      provider.makeAuthorizedRequest(accessToken, integrationURL, HttpMethods.PUT, content);
+
+    } catch (OAuth1HttpRequestException e) {
+      if (e.getCode() == HttpStatus.NOT_FOUND.value()) {
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/services/UserAssignService.java]: Duplicated code. Throws a common exception and deal with this on the exception handler.

File: src/main/java/org/symphonyoss/integration/jira/services/UserAssignService.java [hunk 37]
```diff
@@ -0,0 +1,95 @@
...
+        ErrorResponse errorResponse = new ErrorResponse();
+        errorResponse.setStatus(HttpStatus.NOT_FOUND.value());
+        errorResponse.setMessage(logMessage.getMessage(ISSUEKEY_NOT_FOUND, issueKey));
+        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(errorResponse);
+      }
+      if (e.getCode() == HttpStatus.BAD_REQUEST.value()) {
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/services/UserAssignService.java]: Throws a common exception and deal with this on the exception handler.

**Review by robson-sanchez — CHANGES_REQUESTED** [review]

File: src/main/java/org/symphonyoss/integration/jira/services/SearchAssignableUsersService.java (hunk 32 above)
Comment (robson-sanchez) [reply src/main/java/org/symphonyoss/integration/jira/services/SearchAssignableUsersService.java:52]: DONE

File: src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java (hunk 20 above)
Comment (robson-sanchez) [reply src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java:404]: DONE

**Review by robson-sanchez — APPROVED** [review]

**Pull request merged**